data/bookkeeping/block.go

```diff
@@ -275,9 +293,13 @@ func (block Block) GenesisHash() crypto.Digest {
 }
 
 // WithSeed returns a copy of the Block with the seed set to s.
-func (block Block) WithSeed(s committee.Seed) Block {
+func (block Block) WithSeed(s committee.Seed, proposer basics.Address) Block {
 	c := block
 	c.BlockHeader.Seed = s
+	if !c.BlockHeader.Proposer.IsZero() {
+		panic("Attempt to re-set the proposer.")
+	}
+	c.BlockHeader.Proposer = proposer
```

Comment:
I'd expect this to be guarded by a consensus params check. Otherwise won't the network disagree about what the "right" block is before the upgrade?

Reply:
I think you're right.  I'll investigate and figure out how to test.  The upgrade itself does scare me a bit on this one.

Reply:
Oh, it's handled in:
```
func (vb ValidatedBlock) WithSeed(s committee.Seed, proposer basics.Address) ValidatedBlock {
	newblock := vb.blk
	newblock.BlockHeader.Seed = s

	if vb.blk.ConsensusProtocol().EnableMining {
		newblock.BlockHeader.Proposer = proposer
	}
```

From what I can tell, the existence of this `WithSeed` is only to help implement a mock validatedBlock in `datatest` package. I find that confusing, and will remove it if I can.

Leaving unresolved though, because I need to figure out a good upgrade test.

Reply:
Oh wow, just making an additional note here about why the upgrade is tricky.

`verifyNewSeed` is used to confirm that the `BlockSeed` is correct.  That seemed like a natural place to also check that the block proposer was properly set (because both depend on examining the `unautheticatedProposal`). But that function uses a few different lookbacks, so it uses a 2 round old consensusParams for checking the seed. We want the current params for deciding if Mining is turned on.

All of which points to enough complexity at upgrade time that we need a test.

Reply:
Wow, I completely agree

Reply:
Resolving this because my original concern was addressed